File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const answer = [];
    const genre_sum = {};
    const genre_song = {};
    
    genres.forEach((genre, index) => {
        genre_sum[genre] = (genre_sum[genre] || 0) + plays[index];
        genre_song[genre] = [
            ...(genre_song[genre] || []),
            {id: index, play: plays[index]}
        ];
    });
    
    const genre_sum_sort = Object.entries(genre_sum).sort((a, b) => b[1] - a[1]);
    
    genre_sum_sort.forEach(item => {
        const two_song = genre_song[item[0]].sort((a, b) => {
            return b.play - a.play;
        }).map(song => song.id).slice(0, 2);
        
        answer.push(...two_song);
    })
    
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict

def solution(genres, plays):
    answer = []
    genre_sum = defaultdict(int)
    genre_song = defaultdict(list)
    
    for index, genre in enumerate(genres):
        genre_sum[genre] += plays[index]
        genre_song[genre].append({"id": index, "play": plays[index]})
        
    genre_sum_sort = sorted(genre_sum.items(), key=lambda x: x[1], reverse=True)
    
    for item in genre_sum_sort:
        two_song = sorted(genre_song[item[0]], key=lambda x: x["play"], reverse=True)[:2]
        answer.extend([song["id"] for song in two_song])
    
    return answer
